### PR TITLE
Fixes #19314 - Notification for subscr. expiring soon

### PR DESCRIPTION
**Pull request by dLobatog**
Depends on https://github.com/Katello/katello/pull/6750 and https://github.com/theforeman/foreman/pull/4240/files (this https://github.com/theforeman/foreman/pull/4240/files#diff-2190144d88f9376ea36f321f8f5af34dR71 allows to set actions with more freedom than the current structure).

screenshot from 2017-04-19 13-37-12

### CODE REVIEW

File: app/services/katello/ui_notifications/pulp/proxy_disk_space.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
+module Katello
```
Comment (ohadlevy) [inline app/services/katello/ui_notifications/pulp/proxy_disk_space.rb]: file does not belongs to this pr.

File: app/services/katello/ui_notifications/subscriptions/expire_soon.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      class ExpireSoon
+        class << self
+          def deliver!
+            subs_expiring_soon.each do |subscription|
+              next if notification_already_exists?(subscription)
+              ::Notification.create!(
+                :subject => subscription,
+                :initiator => User.anonymous_admin,
+                :audience => Notification::AUDIENCE_ADMIN,
+                :message => ::UINotifications::StringParser.new(
```
Comment (ohadlevy) [inline app/services/katello/ui_notifications/subscriptions/expire_soon.rb]: if you have the expiry date of the subscription, maybe we should use it for the notification expiry date too?

File: app/services/katello/ui_notifications/subscriptions/expire_soon.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
+module Katello
+  module UINotifications
+    module Subscriptions
+      class ExpireSoon
+        class << self
+          def deliver!
+            subs_expiring_soon.each do |subscription|
+              next if notification_already_exists?(subscription)
```
Comment (ohadlevy) [inline app/services/katello/ui_notifications/subscriptions/expire_soon.rb]: see comments on #6750 around these lines.

File: app/services/katello/ui_notifications/subscriptions/expire_soon.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+              ::Notification.create!(
+                :subject => subscription,
+                :initiator => User.anonymous_admin,
+                :audience => Notification::AUDIENCE_ADMIN,
```
Comment (ohadlevy) [inline app/services/katello/ui_notifications/subscriptions/expire_soon.rb]: IMHO this should be the taxonomy users/admins?

File: app/services/katello/ui_notifications/subscriptions/expire_soon.rb [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+                  ]
+                },
+                :notification_blueprint => blueprint
+              )
+            end
+          end
+
+          def subs_expiring_soon
+            result = []
+            Katello::Subscription.unscoped.all.each do |subscription|
```
Comment (ohadlevy) [inline app/services/katello/ui_notifications/subscriptions/expire_soon.rb]: is there no better query than go over all? at least I would use select instead of the next statement below.

File: db/seeds.d/109-katello-notification-blueprints.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
+blueprints = [
+  {
+    group: _('Proxies'),
+    name: 'pulp_low_disk_space',
+    message: _("%{subject}'s disk is %{percentage} full. Since this proxy is running Pulp, it needs disk space to publish content views. Please ensure the disk does not get full."),
```
Comment (ohadlevy) [inline db/seeds.d/109-katello-notification-blueprints.rb]: not related to this pr.

File: lib/katello/tasks/proxy_disk_space_notification.rake [hunk 7]
```diff
@@ -0,0 +1,13 @@
+namespace :proxy_disk_space do
+  desc <<-END_DESC
+Check the proxy disk space based on Pulp API results and create notifications if
```
Comment (ohadlevy) [inline lib/katello/tasks/proxy_disk_space_notification.rake]: ditto

File: lib/katello/tasks/subs_expiry_notification.rake [hunk 8]
```diff
@@ -0,0 +1,13 @@
+namespace :subscriptions do
+  desc <<-END_DESC
+Check the proxy disk space based on Pulp API results and create notifications if
```
Comment (ohadlevy) [inline lib/katello/tasks/subs_expiry_notification.rake]: wrong description titles etc.

File: app/services/katello/ui_notifications/subscriptions/expire_soon.rb (hunk 4 above)
Comment (dLobatog) [inline app/services/katello/ui_notifications/subscriptions/expire_soon.rb]: I agree, but how do you set this? `AUDIENCE_TAXONOMY` is `subject.user_ids` - if `subject` is the subscription, it will not work. The `subject` can't be the taxonomy

**Comment by dLobatog** [comment]
Updated so that:

- Notifications are expired according to the subscription expiration date
- Finding expiring_soon subs is simplified by using `.select`
- Fixed the typos in the rake task description
- Removed the Pulp low disk space notification commit (https://github.com/Katello/katello/pull/6750/files) from here.

cc @ohadlevy thanks for taking a look before

**Comment by jlsherrill** [comment]
I'm getting an error when running the rake task, i think because Subscription is not taxable:

```
NoMethodError: undefined method `user_ids' for #<Katello::Subscription:0x00000003164bc8>
/home/vagrant/.rvm/gems/ruby-2.2.4/gems/activemodel-4.2.8/lib/active_model/attribute_methods.rb:433:in `method_missing'
/home/vagrant/git/foreman/app/models/notification.rb:43:in `subscriber_ids'
/home/vagrant/git/foreman/app/models/notification.rb:65:in `set_notification_recipients'
/home/vagrant/.rvm/gems/ruby-2.2.4/gems/activesupport-4.2.8/lib/active_support/callbacks.rb:432:in `block in make_lambda'
```

**Comment by akofink** [comment]
@dLobatog any update here?

**Comment by ares** [comment]
it seems rubocop is failing, could someone with permissions set WoC? since we don't have the output in jenkins anymore, I'm retriggering tests [test]

**Comment by ares** [comment]
this needs rebase again, ActiveJob is now in core too, but maybe it's better to wait until https://github.com/Katello/katello/pull/6750 gets in

**Comment by ares** [comment]
LGTM, it would be great if someone from katello could take a look

File: app/services/katello/ui_notifications/subscriptions/expire_soon.rb [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+      class ExpireSoon
+        class << self
+          def deliver!
+            subs_expiring_soon.each do |subscription|
+              next if notification_already_exists?(subscription) &&
+                update_expiration_date(subscription).zero?
+              ::Notification.create!(
+                :subject => subscription,
+                :initiator => User.anonymous_admin,
+                :audience => Notification::AUDIENCE_TAXONOMY,
```
Comment (jlsherrill) [inline app/services/katello/ui_notifications/subscriptions/expire_soon.rb]: Does this require some update to foreman?  I'm getting an error that this constant isn't defined:

```
NameError: uninitialized constant Notification::AUDIENCE_TAXONOMY
	from /home/vagrant/git/foreman_hooks/lib/foreman_hooks/as_dependencies_hook.rb:4:in `load_missing_constant'
	from /home/vagrant/git/katello/app/services/katello/ui_notifications/subscriptions/expire_soon.rb:13:in `block in deliver!'
	from /home/vagrant/git/katello/app/services/katello/ui_notifications/subscriptions/expire_soon.rb:7:in `each'
	from /home/vagrant/git/katello/app/services/katello/ui_notifications/subscriptions/expire_soon.rb:7:in `deliver!'
```

Comment (ares) [reply app/services/katello/ui_notifications/subscriptions/expire_soon.rb]: yeah, there's no such audience, I guess we need to find administrators of a given org for which there's no good option atm, @jlsherrill can we find a user who uploaded the manifest for the organization somehow?

Comment (jlsherrill) [reply app/services/katello/ui_notifications/subscriptions/expire_soon.rb]: We don't track who has uploaded or refreshed a manifest.  I would suggest notifying anyone who has access to do that for the given org via the 'import_manifest' permission

**Comment by ares** [comment]
I think it's not easy to find easily such recipients. There are no relations like this, especially if you consider filtered permissions, where the assignment might be scoped e.g. with "name ~ my". Perhaps we could only search for user who don't have such permission filtered.

**Comment by ares** [comment]
@dLobatog I don't think you need to write custom logic, it's fine to ignore filtering conditions, but I believe it should call `user.can?(permission)` instead of `user.allowed_to?(controller_action)`, the first does not care about specific path that might change but check for permission assignment

**Comment by ares** [comment]
@dLobatog the allowed_to example that you used only verified that user has permission to access "katello/api/v2/subscriptions/import" route based on mapping in lib/katello/permissions/. The fact you pass `organization_id` does not matter, parameters are ignored by `allowed_to` completely. In fact the result is equivalent to `user.can?(:create_subscription)` which returns true if there's any filter with the this permission assigned to the user. Even better, `can?` does not have to parse the list of routes and permissions. Using `allowed_to?` does not make sense since we are really interested in the permission, not whether user has access to specific route. Imagine there would be two routes how you can import manifest...

**Comment by dLobatog** [comment]
@ares Actually I replied w/o looking at the code, `can?` also takes in a subject as an optiional argument, so I'll move to that. Thanks!

**Comment by ares** [comment]
:+1: on the searching of recipients

File: app/services/katello/ui_notifications/subscriptions/expire_soon.rb [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+          def notification_already_exists?(subject)
+            subs_expiration_notification = Notification.unscoped.find_by(:subject => subject)
+            return false if subs_expiration_notification.blank?
+            subs_expiration_notification.notification_blueprint == blueprint
+          end
+
+          def update_expiration_date(subject)
+            expiration_date = subject.pools.
+              find { |pool| pool.expiring_soon? }.end_date.to_date
```
Comment (jlsherrill) [inline app/services/katello/ui_notifications/subscriptions/expire_soon.rb]: this can return nil if none of the pools return true on the block (i.e. no pool is found), and this causes an exception (undefined method end_date for nil class)

File: app/models/katello/subscription.rb [hunk 11]
```diff
@@ -53,5 +55,11 @@ def virt_who?
     def self.humanize_class_name(_name = nil)
       _("Subscription")
     end
+
+    def notification_recipients_ids
+      User.unscoped.all.find_all do |user|
+        user.can?(:import_manifest, organization_id)
```
Comment (jlsherrill) [inline app/models/katello/subscription.rb]: i got an error with this, it seemed that an organization object needed to be passed in, not its id.

**Comment by jlsherrill** [comment]
One thing to note.  I tested this in an org with 78 subscriptions most all with the same expire date.  This results in 78 notifications.  Our current notification UI does not handle lots of notifications very well and requires clicking once per notification to dismiss it.  Would it make more sense to trigger one notification per organization?

**Comment by beav** [comment]
setting waiting on contributor due to conflicts

**Comment by dLobatog** [comment]
[test katello]

**Comment by ares** [comment]
[test katello]

**Comment by jyejare** [comment]
@dLobatog, This needs rebasing.

**Comment by jyejare** [comment]
@dLobatog , Does this PR helps to show an expiration notification before 120 days? Does the notification depend upon the place where the foreman is being accessed (e.g for me, its IST time) ?

**Comment by dLobatog** [comment]
@jyejare Yes, you may change the 120 to any other value in `app/models/katello/pool` (DAYS_EXPIRING_SOON)  although it's not meant to be touched by users. But it can be useful when testing. You can also run `Katello::UINotifications::Subscriptions::ExpireSoon.deliver!` to ensure it runs the check - otherwise it'll only run every 12 hours.

`expiring_soon` is an old issue, it should be checking `Date.today.utc` https://github.com/Katello/katello/blob/master/app/models/katello/pool.rb#L51

**Comment by dLobatog** [comment]
[test katello]

**Comment by dLobatog** [comment]
https://github.com/Katello/katello/pull/7234

**Comment by dLobatog** [comment]
I was missing - https://github.com/Katello/katello/pull/7228/files on my branch, rebasing & pushing again..

**Comment by jyejare** [comment]
@dLobatog , @ares , I tried accessing notifications via org-admin user and via another admin, but I cant ! Looks like there is some issue with users.

File: app/models/katello/concerns/organization_extensions.rb [hunk 12]
```diff
@@ -182,6 +182,17 @@ def cancel_repo_discovery
...
           ::Katello::Resources::Candlepin::Owner.generate_ueber_cert(self.label)
         end
+
+        def expiring_subscriptions
+          subscriptions.select(&:expiring_soon?)
+        end
+
+        def notification_recipients_ids
+          users = User.unscoped.all.find_all do |user|
+            user.can?(:import_manifest) && user.can?(:view_organizations, self)
```
Comment (dLobatog) [inline app/models/katello/concerns/organization_extensions.rb]: @jyejare The problem I had was that previously I was calling user.can?(:import_manifest, self) here. This would not work, because import_manifest is a permission for Subscription, not for Organization. Hence, only admins would return 'true' for that.

tl;dr - It should be fine now for org admins

**Comment by jyejare** [comment]
Tests Performed:

```
1. Notification should be displayed for expiring subscriptions within given days in settings.
2. Notification should only be shown for expiring subscriptions within given days in settings.
3. Notification should not be shown for subscriptions not expiring within given days in settings.
4. Notification body text should be correct.
5. Notification should be shown for org-admin users as well.
6. No duplication of notifications on multiple time delivers.
7. Pulp Notification for disk storage full with the percent of disk full.
```

QE ACK.
